File: dart/test/punctuation_light_zero_arg_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:test/test.dart';

Map<String, Object?> _contract() {
  return (jsonDecode(
            File(
              '../capability_conformance/punctuation_light_zero_arg_contract.json',
            ).readAsStringSync(),
          )
          as Map)
      .cast<String, Object?>();
}

CompiledSpec _compileSource(String source) {
  final parsed = parseSpecWithStagedUserFunctionDefinitions(source);
  validateSpec(parsed);
  return compileSpec(parsed);
}

Object? _semanticAst(Object? value) {
  if (value is List) {
    return [for (final item in value) _semanticAst(item)];
  }
  if (value is Map) {
    return {
      for (final entry in value.entries)
        if (!const {
          'source',
          'source_span',
          'source_method',
          'body_source_span',
          'trailing_block_source_span',
        }.contains(entry.key))
          entry.key: _semanticAst(entry.value),
    };
  }
  return value;
}

String _actionSource(String action) {
  return 'Top::\n'
      ' /x/ -> Done { values = ["a", "b"]; $action }\n'
      'Done::\n'
      ' /x/\n';
}

void main() {
  test('standalone aliases share parenthesized typed ActionIR', () {
    final contract = _contract();
    expect(contract['contract_id'], 'linkedspec-punctuation-light-zero-arg-v1');

    final cases = (contract['standalone_cases']! as List)
        .cast<Map<Object?, Object?>>();
    for (final rawCase in cases) {
      final testCase = rawCase.cast<String, Object?>();
      final bare = parseActionStatement(testCase['bare']! as String).expr;
      final parenthesized = parseActionStatement(
        testCase['parenthesized']! as String,
      ).expr;
      expect(
        _semanticAst(bare.toJson()),
        _semanticAst(parenthesized.toJson()),
        reason: testCase['id']! as String,
      );
      expect(
        bare.kind,
        ((testCase['expected_ast']! as Map<Object?, Object?>)['kind']!
            as String),
        reason: testCase['id']! as String,
      );
    }

    final valuePosition =
        parseActionExpression('return(next)') as ActionCallExpr;
    expect(valuePosition.args.single.value, isA<ActionVariableExpr>());
    expect(
      (valuePosition.args.single.value as ActionVariableExpr).name,
      'next',
    );
    expect(parseActionExpression('next'), isA<ActionVariableExpr>());
  });

  test('terminal receiver aliases share parenthesized typed ActionIR', () {
    final cases = (_contract()['receiver_cases']! as List)
        .cast<Map<Object?, Object?>>();
    for (final rawCase in cases) {
      final testCase = rawCase.cast<String, Object?>();
      final bare = parseActionExpression(testCase['bare']! as String);
      final parenthesized = parseActionExpression(
        testCase['parenthesized']! as String,
      );
      expect(bare, isA<ActionFluentChainExpr>(), reason: '${testCase['id']}');
      expect(
        _semanticAst(bare.toJson()),
        _semanticAst(parenthesized.toJson()),
        reason: testCase['id']! as String,
      );
      final chain = bare as ActionFluentChainExpr;
      expect(chain.calls.last.args, isEmpty, reason: '${testCase['id']}');
    }
  });

  test('retained identifiers and excluded forms do not broaden', () {
    final contract = _contract();
    final retained = (contract['retained_noncall_cases']! as List)
        .cast<Map<Object?, Object?>>();
    for (final rawCase in retained) {
      final testCase = rawCase.cast<String, Object?>();
      final source = testCase['source']! as String;
      final expr = parseActionExpression(source);
      expect(expr, isA<ActionVariableExpr>(), reason: '${testCase['id']}');
      expect((expr as ActionVariableExpr).name, source);
    }

    final invalid = (contract['invalid_syntax_cases']! as List)
        .cast<Map<Object?, Object?>>();
    for (final rawCase in invalid) {
      final testCase = rawCase.cast<String, Object?>();
      final id = testCase['id']! as String;
      final expr = parseActionExpression(testCase['source']! as String);
      expect(expr, isA<ActionRawExpr>(), reason: id);
      expect(
        (expr as ActionRawExpr).reason,
        id == 'intermediate_generic_receiver' ||
                id == 'receiver_trailing_block_without_call'
            ? 'invalid_fluent_chain'
            : 'unsupported_expression',
        reason: id,
      );
    }
  });

  test('terminal alias preserves existing Dart method resolution', () {
    final count = LinkedSpecRuntimeEngine(
      _compileSource(_actionSource('return(values.count)')),
    ).parse('xx');
    expect(count.value, 2);

    final bare = LinkedSpecRuntimeEngine(
      _compileSource(_actionSource('return(values.contains)')),
    ).parse('xx');
    final parenthesized = LinkedSpecRuntimeEngine(
      _compileSource(_actionSource('return(values.contains())')),
    ).parse('xx');
    expect(bare.value, parenthesized.value);
    // Dart already returns 0 for the missing contains needle. Backlog .5 owns
    // that helper-arity drift; this syntax leaf preserves the existing twin.
    expect(bare.value, 0);
  });

  test('neutral fixture matches native and generated Dart paths', () {
    final fixture = (_contract()['future_fixture']! as Map)
        .cast<String, Object?>();
    final source = fixture['spec_source']! as String;
    final input = fixture['input']! as String;
    final expected = fixture['expected'];
    final compiled = _compileSource(source);

    expect(LinkedSpecRuntimeEngine(compiled).parse(input).value, expected);

    final plan = buildGeneratedRulePlan(compiled);
    expect(
      executeGeneratedParserV1(
        compiled,
        plan,
        input,
        'punctuation-light-zero-arg.spec',
      ),
      expected,
    );

    final generated = emitDartSourceV1(
      compiled,
      'punctuation-light-zero-arg.spec',
    );
    expect(generated, contains('linkedspec-generated-source-v1'));
    final encoded = RegExp(
      "const _compiledSpecJsonBase64 = '([^']+)';",
    ).firstMatch(generated)![1]!;
    final normalized = (jsonDecode(utf8.decode(base64Decode(encoded))) as Map)
        .cast<String, Object?>();
    final reconstructed = compileSpec(SpecFile.fromJson(normalized));
    expect(LinkedSpecRuntimeEngine(reconstructed).parse(input).value, expected);
  });
}

File: julia/test/punctuation_light_zero_arg_contract_test.jl
const PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT = JSON3.read(
    read(
        joinpath(
            REPO_ROOT,
            "capability_conformance",
            "punctuation_light_zero_arg_contract.json",
        ),
        String,
    ),
    Dict{String,Any},
)

const _PUNCTUATION_SOURCE_FIELDS = Set([
    "source",
    "source_span",
    "source_method",
    "body_source_span",
    "trailing_block_source_span",
])

function _punctuation_semantic_ast(value)
    if value isa AbstractVector
        return [_punctuation_semantic_ast(item) for item in value]
    elseif value isa AbstractDict
        return Dict(
            String(key) => _punctuation_semantic_ast(item)
            for (key, item) in pairs(value)
            if !(String(key) in _PUNCTUATION_SOURCE_FIELDS)
        )
    end
    return value
end

function _compile_punctuation_source(source::AbstractString)
    spec = parse_spec_with_staged_user_function_definitions(source)
    validate_spec(spec)
    return compile_spec(spec)
end

function _punctuation_action_source(action::AbstractString)
    return "Top::\n" *
           " /x/ -> Done { values = [\"a\", \"b\"]; $action }\n" *
           "Done::\n" *
           " /x/\n"
end

@testset "Julia punctuation-light zero-argument contract" begin
    @test PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["contract_id"] ==
          "linkedspec-punctuation-light-zero-arg-v1"

    @testset "standalone aliases share parenthesized typed ActionIR" begin
        for case in PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["standalone_cases"]
            bare = parse_action_statement(case["bare"]).expr
            parenthesized = parse_action_statement(case["parenthesized"]).expr
            @test _punctuation_semantic_ast(to_json(bare)) ==
                  _punctuation_semantic_ast(to_json(parenthesized))
            @test bare.kind == case["expected_ast"]["kind"]
        end

        value_position = parse_action_expression("return(next)")
        @test value_position isa ActionCallExpr
        @test only(value_position.args).value isa ActionVariableExpr
        @test only(value_position.args).value.name == "next"
        @test parse_action_expression("next") isa ActionVariableExpr
    end

    @testset "terminal receiver aliases share parenthesized typed ActionIR" begin
        for case in PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["receiver_cases"]
            bare = parse_action_expression(case["bare"])
            parenthesized = parse_action_expression(case["parenthesized"])
            @test bare isa ActionFluentChainExpr
            @test _punctuation_semantic_ast(to_json(bare)) ==
                  _punctuation_semantic_ast(to_json(parenthesized))
            @test isempty(last(bare.calls).args)
        end
    end

    @testset "retained identifiers and excluded forms do not broaden" begin
        for case in PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["retained_noncall_cases"]
            expr = parse_action_expression(case["source"])
            @test expr isa ActionVariableExpr
            @test expr.name == case["source"]
        end

        for case in PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["invalid_syntax_cases"]
            expr = parse_action_expression(case["source"])
            @test expr isa ActionRawExpr
            @test expr.reason == if case["id"] in (
                "intermediate_generic_receiver",
                "receiver_trailing_block_without_call",
            )
                "invalid_fluent_chain"
            else
                "unsupported_expression"
            end
        end
    end

    @testset "terminal alias preserves existing Julia method resolution" begin
        count_result = runtime_execute(
            LinkedSpecRuntimeEngine(
                _compile_punctuation_source(
                    _punctuation_action_source("return(values.count)"),
                ),
            ),
            "xx",
        )
        @test count_result.value == 2

        bare_result = runtime_execute(
            LinkedSpecRuntimeEngine(
                _compile_punctuation_source(
                    _punctuation_action_source("return(values.contains)"),
                ),
            ),
            "xx",
        )
        parenthesized_result = runtime_execute(
            LinkedSpecRuntimeEngine(
                _compile_punctuation_source(
                    _punctuation_action_source("return(values.contains())"),
                ),
            ),
            "xx",
        )
        @test bare_result.value == parenthesized_result.value
        # Julia already returns 0 for the missing contains needle. Backlog .5
        # owns that helper-arity drift; this syntax leaf preserves the twin.
        @test bare_result.value == 0
    end

    @testset "neutral fixture matches native and generated Julia paths" begin
        fixture = PUNCTUATION_LIGHT_ZERO_ARG_CONTRACT["future_fixture"]
        compiled = _compile_punctuation_source(fixture["spec_source"])
        expected = JSON3.read(JSON3.write(fixture["expected"]), Dict{String,Any})

        @test runtime_execute(
            LinkedSpecRuntimeEngine(compiled),
            fixture["input"],
        ).value == expected

        plan = build_generated_rule_plan(compiled)
        @test execute_generated_parser_v2(
            compiled,
            plan,
            fixture["input"],
            "punctuation-light-zero-arg.spec",
        ) == expected

        generated = emit_julia_source_v2(
            compiled,
            "punctuation-light-zero-arg.spec",
        )
        @test occursin("linkedspec-generated-source-v2", generated)
        encoded = match(
            r"const _COMPILED_SPEC_JSON_HEX = \"([0-9a-f]+)\"",
            generated,
        )
        @test encoded !== nothing
        normalized = JSON3.read(String(hex2bytes(encoded.captures[1])))
        reconstructed = compile_spec(from_json(SpecFile, normalized))
        @test runtime_execute(
            LinkedSpecRuntimeEngine(reconstructed),
            fixture["input"],
        ).value == expected
    end
end
